File: components/stepper-rp2040-firmware/src/board_V3_arduino/src/stepper.cpp
#include "stepper.h"
#include "stepper_pio.h"
#include "rgb_led.h"
#include <hardware/clocks.h>
#include <hardware/irq.h>

// Map from ms index -> (label, reduction, ms1, ms2, SG_adjustment, serialport node address).
struct MicrostepEntry {
    const char* label;
    float reduction;
    uint8_t ms1;
    uint8_t ms2;
    uint8_t sg_adj;
    uint8_t sp_na;
};

static const MicrostepEntry MICROSTEP_MAP[] = {
    { "1/8",  0.125f,    0, 0, 1, 0 },
    { "1/16", 0.0625f,   1, 1, 2, 3 },
    { "1/32", 0.03125f,  1, 0, 4, 1 },
    { "1/64", 0.015625f, 0, 1, 8, 2 },
};

Stepper* Stepper::_instance = nullptr;

Stepper::Stepper(uint32_t max_frequency, uint32_t frequency, bool debug)
    : _tmc(Serial1, UART_RX, UART_TX, /*mtr_id placeholder*/ 0, 230400),
      _debug(debug),
      _max_frequency(max_frequency),
      _frequency(frequency),
      // Adafruit_NeoPixel grabs a state machine on pio0; park the stepper on
      // pio1 so there is no contention for programs or SMs.
      _pio(pio1),
      _sm0(0), _sm1(1), _sm2(2),
      _offset0(0), _offset1(0), _offset2(0),
      _full_rev(1600),
      _SG_adj(1),
      _sp_na(0),
      _max_homing_revs(5),
      _max_steps(8000),
      _stepper_spinning(false),
      _stallguarded(true)
{
    Serial.println("\nUploading stepper_controller ...");
    _instance = this;

    // Output pins.
    pinMode(DIR_PIN, OUTPUT);
    pinMode(MS1_PIN, OUTPUT);
    pinMode(MS2_PIN, OUTPUT);

    // Load the three PIO programs.
    _offset0 = pio_add_program(_pio, &steps_mot_program);
    _offset1 = pio_add_program(_pio, &stop_stepper_program);
    _offset2 = pio_add_program(_pio, &steps_counter_program);

    // Configure the step pin for PIO output.
    pio_gpio_init(_pio, STEP_PIN);
    pio_sm_set_consecutive_pindirs(_pio, _sm0, STEP_PIN, 1, true);

    // SM0: frequency generator driving STEP_PIN.
    {
        pio_sm_config c = pio_get_default_sm_config();
        sm_config_set_wrap(&c, _offset0, _offset0 + steps_mot_program.length - 1);
        sm_config_set_set_pins(&c, STEP_PIN, 1);
        float div = (float)clock_get_hz(clk_sys) / (float)_frequency;
        sm_config_set_clkdiv(&c, div);
        pio_sm_init(_pio, _sm0, _offset0, &c);
        // Prime the FIFO with the slowest-speed delay. The PIO's `pull noblock`
        // picks this up on the first iteration after the SM is enabled.
        pio_sm_put_blocking(_pio, _sm0, 65535);
        pio_sm_set_enabled(_pio, _sm0, false);
    }

    // SM1: stop-on-count watcher. Raises PIO IRQ 1 when X underflows.
    {
        pio_sm_config c = pio_get_default_sm_config();
        sm_config_set_wrap(&c, _offset1, _offset1 + stop_stepper_program.length - 1);
        sm_config_set_in_pins(&c, STEP_PIN);
        float div = (float)clock_get_hz(clk_sys) / (float)_max_frequency;
        sm_config_set_clkdiv(&c, div);
        pio_sm_init(_pio, _sm1, _offset1, &c);
        pio_sm_set_enabled(_pio, _sm1, false);
    }

    // Route PIO IRQ 1 (raised by SM1's `irq rel 0`) to NVIC PIO1_IRQ_0.
    pio_set_irq0_source_enabled(_pio, pis_interrupt1, true);
    irq_set_exclusive_handler(PIO1_IRQ_0, Stepper::onPioIrq);
    irq_set_enabled(PIO1_IRQ_0, true);

    // SM2: step counter.
    {
        pio_sm_config c = pio_get_default_sm_config();
        sm_config_set_wrap(&c, _offset2, _offset2 + steps_counter_program.length - 1);
        sm_config_set_in_pins(&c, STEP_PIN);
        float div = (float)clock_get_hz(clk_sys) / (float)_max_frequency;
        sm_config_set_clkdiv(&c, div);
        pio_sm_init(_pio, _sm2, _offset2, &c);
        // Enable before setPlsCounter(): exec_wait_blocking only completes on
        // an enabled SM (the forced instruction pre-empts any stalled wait,
        // but the SM must be clocked to actually execute it).
        pio_sm_set_enabled(_pio, _sm2, true);
        setPlsCounter(0);
    }

    // Pick the microstep setting first: it selects the TMC UART node address.
    if (!microStep(0)) {
        Serial.println("ERROR: invalid microstep selection");
        return;
    }

    // The UART was opened with mtr_id=0 at construction; now that the
    // microstep selection told us the real node address, point the driver at it.
    _tmc.setMotorId(_sp_na);

    if (_tmc.test()) {
        setStallguard(0);
    }

    _max_steps = _max_homing_revs * _full_rev;
}

void Stepper::onPioIrq() {
    if (!_instance) return;
    if (pio1->irq & (1u << 1)) {
        pio1->irq = (1u << 1);  // clear
        pio_sm_set_enabled(pio1, _instance->_sm0, false);
        _instance->_stepper_spinning = false;
    }
}

void Stepper::onStallguard() {
    if (!_instance) return;
    pio_sm_set_enabled(pio1, _instance->_sm0, false);
    _instance->_stallguarded = true;
    _instance->_stepper_spinning = false;
}

bool Stepper::tmcTest() {
    return _tmc.test();
}

void Stepper::stopStepper() {
    pio_sm_set_enabled(_pio, _sm0, false);
    _stepper_spinning = false;
}

void Stepper::startStepper() {
    _stepper_spinning = true;
    pio_sm_set_enabled(_pio, _sm0, true);
}

void Stepper::deactivatePio() {
    pio_sm_set_enabled(_pio, _sm0, false);
    pio_sm_set_enabled(_pio, _sm1, false);
    pio_sm_set_enabled(_pio, _sm2, false);
    pio_remove_program(_pio, &steps_mot_program, _offset0);
    pio_remove_program(_pio, &stop_stepper_program, _offset1);
    pio_remove_program(_pio, &steps_counter_program, _offset2);
    Serial.println("State Machines deactivated");
}

// Back-to-back pio_sm_exec writes stomp each other: the second SMn_INSTR write
// can arrive before the SM has latched the first. MicroPython gets away with
// consecutive sm.exec() calls only because the interpreter inserts a few µs
// between them. A 1 µs gap here is ~125 PIO cycles at 125 MHz, plenty.
static inline void exec_pair(PIO pio, uint sm, uint instr_a, uint instr_b) {
    pio_sm_exec(pio, sm, instr_a);
    delayMicroseconds(1);
    pio_sm_exec(pio, sm, instr_b);
    delayMicroseconds(1);
}

void Stepper::setPlsToDo(uint32_t val) {
    pio_sm_set_enabled(_pio, _sm1, true);
    pio_sm_put_blocking(_pio, _sm1, val);
    exec_pair(_pio, _sm1,
              pio_encode_pull(false, false),
              pio_encode_mov(pio_x, pio_osr));
}

void Stepper::setPlsCounter(uint32_t val) {
    pio_sm_put_blocking(_pio, _sm2, val);
    exec_pair(_pio, _sm2,
              pio_encode_pull(false, false),
              pio_encode_mov(pio_x, pio_osr));
}

int32_t Stepper::getPlsCount() {
    exec_pair(_pio, _sm2,
              pio_encode_mov(pio_isr, pio_x),
              pio_encode_push(false, false));
    if (!pio_sm_is_rx_fifo_empty(_pio, _sm2)) {
        uint32_t v = pio_sm_get(_pio, _sm2);
        return (int32_t)(-(int32_t)v);
    }
    return -1;
}

void Stepper::setStallguard(int threshold) {
    if (threshold < 0) threshold = 0;
    if (threshold > 255) threshold = 255;
    _tmc.setStallguardCallback((uint8_t)threshold, Stepper::onStallguard);
    if (_debug) {
        if (threshold != 0) {
            Serial.print("Setting StallGuard (irq to GPIO 11) to value ");
            Serial.println(threshold);
        } else {
            Serial.println("Setting StallGuard (irq to GPIO 11) to 0, meaning max possible torque.");
        }
    }
}

uint32_t Stepper::getStepperValue(uint32_t stepper_freq) {
    if (stepper_freq == 0) return 0;
    int64_t v = ((int64_t)_frequency - (int64_t)stepper_freq * (int64_t)PIO_FIX) /
                ((int64_t)stepper_freq * (int64_t)PIO_VAR);
    if (v < 0) v = 0;
    if (v > 0xFFFFFFFFLL) v = 0xFFFFFFFFLL;
    return (uint32_t)v;
}

bool Stepper::microStep(int ms) {
    if (ms < 0 || ms >= (int)(sizeof(MICROSTEP_MAP) / sizeof(MICROSTEP_MAP[0]))) {
        Serial.println("Wrong parameter for micro_step");
        return false;
    }
    const MicrostepEntry& e = MICROSTEP_MAP[ms];
    digitalWrite(MS1_PIN, e.ms1);
    digitalWrite(MS2_PIN, e.ms2);
    Serial.print("Microstepping set to ");
    Serial.println(e.label);

    _full_rev = (uint32_t)(STEPPER_STEPS / e.reduction);
    _SG_adj = e.sg_adj;
    _sp_na = e.sp_na;
    Serial.print("Full revolution takes ");
    Serial.print(_full_rev);
    Serial.println(" steps");
    return true;
}

bool Stepper::homing(uint32_t h_speed, uint32_t stepper_freq, int startup_loops,
                     uint32_t retract_time_ms, int32_t retract_steps) {
    setStallguard(0);
    uint32_t max_homing_ms = retract_time_ms +
        (uint32_t)(_max_homing_revs * 1000ULL * _full_rev / stepper_freq);

    bool do_once = true;
    _stallguarded = false;
    stopStepper();
    setPlsCounter(0);
    setPlsToDo((uint32_t)retract_steps + _max_steps);
    pio_sm_put_blocking(_pio, _sm0, h_speed);
    startStepper();

    uint32_t min_sg_expected = (uint32_t)(0.15f * (float)stepper_freq / (float)_SG_adj);
    uint32_t sg_threshold = (uint32_t)(0.8f * (float)min_sg_expected);
    uint32_t sg_threshold_diag = (uint32_t)(0.45f * (float)min_sg_expected);

    if (_debug) {
        Serial.print("Homing with stepper speed of ");
        Serial.print(stepper_freq);
        Serial.print("Hz and UART StallGuard threshold of ");
        Serial.println(sg_threshold);
    }

    uint32_t t_ref = millis();
    uint32_t last_tick = t_ref;
    int32_t min_sg_seen = 1024;
    int i = 0;
    while (millis() - t_ref < max_homing_ms) {
        int32_t sg = _tmc.getStallguardResult();
        if (sg < min_sg_seen) min_sg_seen = sg;
        i++;
        // Diagnostic heartbeat once per second so a hang vs. a slow timeout
        // is distinguishable from the serial log alone.
        uint32_t now = millis();
        if (_debug && (now - last_tick) >= 1000) {
            last_tick = now;
            Serial.print("  .. t=");
            Serial.print(now - t_ref);
            Serial.print("ms i=");
            Serial.print(i);
            Serial.print(" sg=");
            Serial.print(sg);
            Serial.print(" min_sg=");
            Serial.print(min_sg_seen);
            Serial.print(" diag=");
            Serial.println(_stallguarded ? 1 : 0);
        }
        if (i > startup_loops) {
            if (do_once) {
                setStallguard((int)sg_threshold_diag);
                do_once = false;
            }
            if ((uint32_t)sg < sg_threshold || _stallguarded) {
                stopStepper();
                int32_t pls = getPlsCount();
                // Matches the Python check: a -1 "couldn't read" compares less
                // than 0.95 * max_steps and is treated as a valid homing.
                if (pls < (int32_t)(0.95f * (float)((uint32_t)retract_steps + _max_steps))) {
                    int times;
                    float time_s;
                    float bright;
                    if (_stallguarded) {
                        times = 1; time_s = 0.01f; bright = 0.8f;
                    } else {
                        times = 3; time_s = 0.05f; bright = 0.1f;
                    }
                    rgb_led.flashColor("red", bright, times, time_s);

                    if (_debug) {
                        if (_stallguarded) Serial.println("StallGuard detections via DIAG pin");
                        Serial.print("Homing reached in ");
                        Serial.print(i);
                        Serial.print(" iterations, ");
                        Serial.print(millis() - t_ref);
                        Serial.println(" ms");
                    }
                    return true;
                }
            }
        }
    }

    setStallguard(0);
    stopStepper();
    Serial.println("Failed homing");
    return false;
}

void Stepper::retract(uint32_t speed, int startup_loops,
                      uint32_t* retract_time_ms, int32_t* retract_steps) {
    setStallguard(0);
    stopStepper();
    setPlsCounter(0);
    setPlsToDo(_max_steps);
    pio_sm_put_blocking(_pio, _sm0, speed);
    startStepper();
    if (_debug) {
        Serial.println("Retract the stepper prior the first home search");
    }
    uint32_t t_ref = millis();
    for (int i = 0; i < startup_loops; i++) {
        (void)_tmc.getStallguardResult();
    }
    stopStepper();
    *retract_time_ms = millis() - t_ref;
    *retract_steps = getPlsCount();
    if (*retract_steps < 0) *retract_steps = 0;
}

bool Stepper::centering(uint32_t stepper_freq) {
    uint32_t min_freq = 400 * _SG_adj;
    uint32_t max_freq = 1200 * _SG_adj;
    stepper_freq = _SG_adj * stepper_freq;
    if (stepper_freq < min_freq) stepper_freq = min_freq;
    if (stepper_freq > max_freq) stepper_freq = max_freq;

    uint32_t stepper_val = getStepperValue(stepper_freq);
    int startup_loops = 10;

    digitalWrite(DIR_PIN, LOW);
    uint32_t retract_time_ms = 0;
    int32_t retract_steps = 0;
    retract(stepper_val, startup_loops, &retract_time_ms, &retract_steps);

    digitalWrite(DIR_PIN, HIGH);
    if (!homing(stepper_val, stepper_freq, startup_loops, retract_time_ms, retract_steps)) {
        stopStepper();
        rgb_led.flashColor("blue", 0.1f, 10, 0.05f);
        return false;
    }

    digitalWrite(DIR_PIN, LOW);
    if (!homing(stepper_val, stepper_freq, startup_loops, retract_time_ms, retract_steps)) {
        stopStepper();
        rgb_led.flashColor("blue", 0.1f, 10, 0.05f);
        return false;
    }

    int32_t steps_range = getPlsCount();
    if (steps_range < 0) steps_range = 0;
    int32_t half_range = steps_range / 2;
    digitalWrite(DIR_PIN, HIGH);
    setPlsToDo((uint32_t)half_range);
    pio_sm_put_blocking(_pio, _sm0, stepper_val);
    startStepper();
    uint32_t centering_time_ms = 100 + (uint32_t)((int64_t)half_range * 1000LL / (int64_t)stepper_freq);
    if (_debug) {
        Serial.print("Counted ");
        Serial.print(steps_range);
        Serial.print(" steps between the 2 homes; centering to ");
        Serial.println(half_range);
    }
    delay(centering_time_ms);
    if (!_stepper_spinning) {
        rgb_led.flashColor("green", 0.2f, 3, 0.05f);
    }
    return true;
}

File: components/stepper-rp2040-firmware/src/board_V2/stepper.py
"""
Andrea Favero 16/05/2025

Micropython code for Raspberry Pi Pico (RP2040 and RP2350)
It demonstrates how to use TMC2209 StallGuard function for stepper sensorless homing.
PIO is used by the RP2040 (or RP2350) to generate the stepper steps frequency.




MIT License

Copyright (c) 2025 Andrea Favero

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from TMC_2209_driver import *
from rgb_led import rgb_led
from rp2 import PIO, StateMachine, asm_pio, asm_pio_encode
from machine import Pin
import time




class Stepper:

    def __init__(self, max_frequency=125000000, frequency=5000000, debug=False):
        print("\nUploading stepper_controller ...")
        
        # debug bool for printout
        self.debug = debug
        
        # frequencies for the PIO state machines
        self.max_frequency = max_frequency
        self.frequency = frequency 
        
        # (RP2040) GPIO pins
        self.STEPPER_STEPS_PIN = Pin(5, Pin.OUT)
        self.STEPPER_DIR = Pin(6, Pin.OUT)
        self.STEPPER_MS1 = Pin(3, Pin.OUT)
        self.STEPPER_MS2 = Pin(4, Pin.OUT)
        self.UART_RX_PIN = Pin(13)
        self.UART_TX_PIN = Pin(12)
        
        # stepper characteristics
        self.STEPPER_STEPS = 200        # number of (full) steps per revolution
        
        # PIO intrustions for steps generation (sm0)
        self.PIO_VAR = 2                # number of PIO commands repeated, when PIO stepper state machine is called
        self.PIO_FIX = 37               # fix number of PIO commands, when PIO stepper state machine is called
        
        # pre-encode some of the PIO instructions (>100 times faster)
        self.PULL_ENCODED = asm_pio_encode("pull()", 0)
        self.MOV_X_OSR_ENCODED = asm_pio_encode("mov(x, osr)", 0)
        self.PUSH_ENCODED = asm_pio_encode("push()", 0)
        self.MOV_ISR_X_ENCODED = asm_pio_encode("mov(isr, x)", 0)        

        # state machine for stepper steps generation
        self.sm0 = StateMachine(0, self.steps_mot_pio, freq=self.frequency, set_base=self.STEPPER_STEPS_PIN)
        self.sm0.put(65535)             # initial OSR to max value (minimum stepper speed)
        self.sm0.active(0)              # state machine is kept deactivated
        
        # state machine for stopping the stepper
        self.sm1 = StateMachine(1, self.stop_stepper_pio, freq=self.max_frequency, in_base=self.STEPPER_STEPS_PIN)
        self.sm1.irq(self._stop_stepper_handler)
        self.sm1.active(0)              # state machine is kept deactivated
        
        # state machine for stepper steps tracking
        self.sm2 = StateMachine(2, self.steps_counter_pio, freq=self.max_frequency, in_base=self.STEPPER_STEPS_PIN)
        self.set_pls_counter(0)         # sets the initial value for stepper steps counting
        self.sm2.active(1)              # starts the state machine for stepper steps counting
        
        # microsteppingmap --> setting: (descriptor, reduction, ms1, ms2)
        self.microstep_map = {0: ("1/8", 0.125, 0, 0),
                              1: ("1/16", 0.0625, 1, 1),
                              2: ("1/32", 0.03125, 1, 0),
                              3: ("1/64", 0.015625, 0, 1)}
        
        # microstep resolution configuration (internal pull-down resistors)
        ms = self.micro_step(0)   # 0=1/8, 1=1/16, 2=1/32, 3=1/64
        
        # get the steps for a full revolution
        if ms is not None:
            self.full_rev = self.get_full_rev(ms)
        
        # instantiation of tmc2209 driver
        # args: pin_step, pin_dir, pin_en, rx_pin, tx_pin, mtr_id=0, baudrate=115200, serialport=0
        self.tmc = TMC_2209( rx_pin=self.UART_RX_PIN, tx_pin=self.UART_TX_PIN, mtr_id=0, serialport=0 )
        
        # set StallGuard to max torque for the GPIO interrupt callback
        self.set_stallguard(threshold = 0)
        
        # flag for stepper spinning status (True when the motor spins)
        self.stepper_spinning = False
        
        # instance variable for the max number of stepper revolution to find home
        self.max_homing_revs = 5
        
        # instance variable for the max number of steps to be done while homing
        self.max_steps = self.max_homing_revs * self.full_rev
        
        # instance variable for the steps to be done
        self.steps_to_do = self.max_steps
        









    # sm0, the PIO program for generating steps 
    @asm_pio(set_init=PIO.OUT_LOW)
    def steps_mot_pio():      # Note: without 'self' in PIO callback function !
        """
        A frequency is generated at PIO: The frequency remains fix (it can activated and stopped).
        A Pin is associated to the Pin
        The speed can be varied by changing the off time (delay) between ON pulses
        This method allows for stepper speed variation, not really for precise positioning.
        """
        label("main")         # entry point from jmp
        pull(noblock)         # get delay data from osr, and maintained in memory
        mov(x,osr)            # OSR (Output Shift Register) content is moved to x, to prepare for future noblock pulls
        mov(y, x)             # delay assigned to y scretch register
        set(pins, 1)  [15]    # set the pin high 16 times
        set(pins, 1)  [15]    # set the pin high 16 times
        label("delay")        # entry point when x_dec>0
        set(pins, 0)          # set the pin low 1 time
        jmp(y_dec, "delay")   # y_dec-=1 if y_dec>0 else jump to "delay" 
        jmp("main")           # jump to "main"


    # sm1, the PIO program for stopping the sm0 (steps generator)
    @asm_pio()
    def stop_stepper_pio():   # Note: without 'self' in PIO callback function !
        """
        This PIO fiunction:
            - Reads the lowering edges from the output GPIO that generates the steps.
            - De-counts the 32bits OSR.
            - The starting value, to decount from, is sent by inline command.
            - Once the decount is zero it stops the steps generation (sm0).
        """
        label("wait_for_step")
        wait(1, pin, 0)       # wait for step signal to go HIGH
        wait(0, pin, 0)       # wait for step signal to go LOW (lowering edge)
        jmp(x_dec, "wait_for_step")
        irq(block, rel(0))    # trigger an IRQ


    # sm2, the PIO program for counting steps
    @asm_pio()
    def steps_counter_pio():   # Note: without 'self' in PIO callback function !
        """
        This PIO fiunction:
            -Reads the rising edges from the output GPIO that generates the steps.
            -De-counts the 32bit OSR.
            -The starting value, to decount from, is sent by inline command.
            -The reading is pushed to the rx_fifo on request, via an inline command.
        """
        label("loop")
        wait(0, pin, 0)       # wait for step signal to go LOW
        wait(1, pin, 0)       # wait for step signal to go HIGH (rising edge)
        jmp(x_dec, "loop")    # decrement x and continue looping


    def set_pls_to_do(self, val):
        """Sets  the initial value for the stpes to do, from which to decount from.
           Pre-encoded instructions are >140 times faster."""
        self.sm1.put(val)                      # val value is passed to sm1 (StateMachine 1)
        self.sm1.exec(self.PULL_ENCODED)       # execute pre-encoded 'pull()' instruction
        self.sm1.exec(self.MOV_X_OSR_ENCODED)  # execute pre-encoded 'mov(x, osr)' instruction
        self.sm1.active(1)                     # sm1 is activated


    def set_pls_counter(self, val):
        """Sets  the initial value for the stpes counter, from which to decount from.
           Pre-encoded instructions are >140 times faster."""
        self.sm2.put(val)
        self.sm2.exec(self.PULL_ENCODED)       # execute pre-encoded 'pull()' instruction
        self.sm2.exec(self.MOV_X_OSR_ENCODED)  # execute pre-encoded 'mov(x, osr)' instruction


    def get_pls_count(self):
        """Gets the steps counter value.
           Pre-encoded instructions are >140 times faster."""
        self.sm2.exec(self.MOV_ISR_X_ENCODED)  # execute pre-encoded 'mov(isr, x)' instruction
        self.sm2.exec(self.PUSH_ENCODED)       # execute pre-encoded 'push()' instruction
        if self.sm2.rx_fifo:                   # case there is data in the rx buffer
            return -self.sm2.get() & 0xffffffff  # return the current sm2 counter value (32-bit unsigned integer)
        else:                                  # case the rx buffer has no data
            return -1                          # returning -1 is a clear exception for a positive counter ...)


    def set_steps_value(self, val):
        """Set the ref position (after homing)."""
        self.position = val


    def _is_tmc_powered(self):
        """Check if the TMC driver is energized, via the UART."""
        rtn = self.read_stallguard()
        if isinstance(rtn, int):
            print("TMC_2209 driver is powered")
            return 1
        elif rtn == ():
            print("TMC_2209 driver is not powered (or UART not connected)")
            return 0


    def read_stallguard(self):
        """Gets the instant StallGuard value from the TMC driver, via the UART."""
        return self.tmc.getStallguard_Result()


    def set_stallguard(self, threshold):
        """Sets the StallGuard threshold at TMC driver, via the UART."""
        # clamp the SG threshold between 0 and 255
        threshold = max(0, min(threshold, 255))
        
        # set the StallGuard threshold and the call-back handler function
        self.tmc.setStallguard_Callback(threshold = threshold, handler = self._stallguard_callback)
        if self.debug:
            if threshold != 0:
                print("\nSetting StallGuard (irq to GPIO) to  value {}.".format(threshold))
            else:
                print("\nSetting StallGuard (irq to GPIO) to 0, meaning max possible torque.")


    def get_stepper_frequency(self, pio_val):
        """Convert the PIO value (delay) to stepper frequency."""
        if pio_val > 0:
            return int(self.frequency / (pio_val * self.PIO_VAR + self.PIO_FIX))
        else:
            return None


    def get_stepper_value(self, stepper_freq):
        """Convert the stepper frequency to PIO value (delay)."""
        if stepper_freq > 0:
            return int((self.frequency - stepper_freq * self.PIO_FIX) / (stepper_freq * self.PIO_VAR))
        else:
            return None


    def micro_step(self, ms):
        """Sets the GPIO for th decided microstepping (ms) setting."""
        settings = self.microstep_map.get(ms)  # retrieve the corresponding settings
        
        if settings:                           # case the ms setting exists
            ms_txt, k, ms1, ms2 = settings     # ms map elements are assigned
            self.STEPPER_MS1.value(ms1)        # STEPPER_MS1 output set to ms1 value
            self.STEPPER_MS2.value(ms2)        # STEPPER_MS2 output set to ms2 value
            print(f"Microstep to {ms_txt}")    # feedback is printed to the terminal
            return ms
        else:                                  # case the ms setting does not exist
            print("Wrong parameter for micro_step") # feedback is printed to the terminal
            return None


    def get_full_rev(self, ms):
        """Calculates the steps for one fulle revolution of the stepper."""
        settings = self.microstep_map.get(ms)  # retrieve the corresponding settings
        
        if settings:                           # case the ms setting exists
            ms_txt, k, ms1, ms2 = settings     # ms map elements are assigned
            full_rev = int(self.STEPPER_STEPS/k)  # steps for a stepper full revolution
            print(f"Full revolution takes {full_rev} steps")  # feedback is printed to the terminal
            return full_rev
        else:                                  # case the ms setting does not exist
            print("Wrong parameter for full revolution") # feedback is printed to the terminal
            return None


    def _stop_stepper_handler(self, sm0):
        """Call-back function by a PIO interrupt"""
        self.sm0.active(0)                     # state machine for stepper-steps generation is deactivated
        self.stepper_spinning = False          # flag tracking the stepper spinning is set False
        if self.steps_to_do < self.max_steps / 2: # case stepper stops within half way of max_steps for homing
            rgb_led.flash_color('green', bright=0.2, times=1, time_s=0.01) # flashing red led


    def _stallguard_callback(self, channel):
        """Call-back function from the StallGuard."""
        print("\nStallGuard detections\n")
        self.sm0.active(0)                     # state machine for stepper-steps generation is deactivated
        self.stepper_spinning = False          # flag tracking the stepper spinning is set False


    def stop_stepper(self):
        self.sm0.active(0)                     # state machine for stepper-steps generation is deactivated
        self.stepper_spinning = False          # flag tracking the stepper spinning is set False


    def start_stepper(self):
        self.stepper_spinning = True           # flag tracking the stepper spinning is set True
        self.sm0.active(1)                     # state machine for stepper-steps generation is activated


    def deactivate_pio(self):
        """Function to deactivate PIO."""
        self.sm0.active(0)                     # sm0 is deactivated
        self.sm1.active(0)                     # sm1 is deactivated
        self.sm2.active(0)                     # sm2 is deactivated
        PIO(0).remove_program()                # reset SM0 block
        PIO(1).remove_program()                # reset SM1 block
        print("State Machines deactivated")


    def _homing(self, h_speed, stepper_freq):
        """
        Spins the stepper at stepper_freq until StillGuard value below threshold.
        Argument h_speed is the PIO value to get stepper_freq at the GPIO pin.
        """
        self.stop_stepper()                      # stepper is stopped (in the case it wasn't)
        self.set_pls_counter(0)                  # sets the initial value for stepper steps counting
        self.set_pls_to_do(self.max_steps)       # (max) number of steps to find home
        self.sm0.put(h_speed)                    # stepper speed
        self.start_stepper()                     # stepper is started

        min_sg_expected = int(0.15 * stepper_freq)  # expected minimum StallGuard value on free spinning
        k = 0.8                                  # reduction coeficient (it should be 0.7 ~ 0.8, tune it on your need)
        sg_threshold = int(k * min_sg_expected)  # StallGuard threshold 
        max_homing_ms = int(self.max_homing_revs * 1000 * self.full_rev / stepper_freq) # timeout in ms
        
        if self.debug:                           # case self.debug is set True
            sg_list = []                         # list for debug purpose
            print(f"Homing with stepper speed of {stepper_freq}Hz and StallGuard threshold of {sg_threshold}")
        
        t_ref = time.ticks_ms()                  # time reference
        i = 0                                    # iterator index
        while time.ticks_ms() - t_ref < max_homing_ms: # while loop until timeout (unless SG detection)
            sg = self.tmc.getStallguard_Result() # StallGuard value is retrieved
            if self.debug:                       # case self.debug is set True
                sg_list.append(sg)               # StallGuard value is appended to the list
            i+=1                                 # iterator index is increased
            if i > 10:                           # case at least 10 SG readings (avoids startup effect to SG)
                if sg < sg_threshold:            # case StallGuard value lower than threshold
                    self.stop_stepper()          # stepper is stopped (if not done by the _homing func)
                    rgb_led.flash_color('red', bright=0.3, times=1, time_s=0.01) # flashing red led
                    if self.debug:               # case self.debug is True
                        print(f"Homing reached. Last SG values: {sg_list}.", \
                              f"Total of {i} iterations in {time.ticks_ms()-t_ref} ms\n")
                    return True                  # return True
                else:                            # case StallGuard value higher than threshold
                    if self.debug:               # case self.debug is set True
                        del sg_list[0]           # first element in list is deleted
                    else:                        # case self.debug is set False
                        pass                     # do nothing
        
        self.stop_stepper()                      # stepper is stopped (if not done by the _homing func)
        print("Failed homing")                   # feedback is printed to the Terminal
        return False                             # False is returned when homing fails



    def centering(self, stepper_freq):
        """
        SENSORLESS homing, on both directions, to stop the stepper in the middle.
        Argument is the stepper speed, in Hz.
        Clockwise (CW) and counter-clockwise (CCW) also depends on motor wiring.
        """
        
        stepper_freq = max(400, min(1200, stepper_freq))   # homing speed (frequency) clamped in range 400 ~ 1200 Hz
        stepper_val = self.get_stepper_value(stepper_freq) # stepper pio value is calculated
        
        self.STEPPER_DIR.value(1)                       # set stepper direction to 1 (CW)
        if self._homing(stepper_val, stepper_freq):     # call the _homing function at CW
            self.STEPPER_DIR.value(0)                   # set stepper direction to 0 (CCW)
            if self._homing(stepper_val, stepper_freq): # call the _homing function at CW
                steps_range = self.get_pls_count()      # retrieves steps done on previous activation run
                half_range = int(steps_range/2)         # half range
                self.STEPPER_DIR.value(1)               # set stepper direction to 1 (CW)
                self.steps_to_do = half_range           # steps to be done assigned to instance variable
                self.set_pls_to_do(self.steps_to_do)    # number of steps to do now is half range
                self.sm0.put(stepper_val)               # stepper speed
                self.start_stepper()                    # stepper is started
                if self.debug:                          # case self.debug is set True
                    print(f"Counted {steps_range} steps in between the 2 homes")
                    print(f"Positioning the stepper at {half_range} from the last detected home")
                return True                             # True is returned when centering succeeds
            
        else:                                           # case the homing fails
            self.steps_to_do = self.max_steps           # max number of homing steps is assigned to the steps to be done next
            self.stop_stepper()                         # stepper is stopped (in the case it wasn't)
            return False                                # False is returned when centering fails


            
        

